File: android/app/src/main/java/com/rajarsheechatterjee/LNReader/TTSHighlightModule.java
package com.rajarsheechatterjee.LNReader;

import android.os.Bundle;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.speech.tts.Voice;
import com.facebook.react.bridge.*;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public class TTSHighlightModule extends ReactContextBaseJavaModule {
    private TextToSpeech tts;
    private ReactContext reactContext;
    private boolean isTtsReady = false;

    public TTSHighlightModule(ReactApplicationContext context) {
        super(context);
        this.reactContext = context;
        initializeTTS();
    }

    @Override
    public String getName() {
        return "TTSHighlight";
    }

    private void initializeTTS() {
        tts = new TextToSpeech(getReactApplicationContext(), status -> {
            if (status == TextToSpeech.SUCCESS) {
                isTtsReady = true;
                setupProgressListener();
            } else {
                sendEvent("onSpeechError", createErrorMap("TTS Initialization failed"));
            }
        });
    }

    private void setupProgressListener() {
        tts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
            @Override
            public void onRangeStart(String utteranceId, int start, int end, int frame) {
                WritableMap params = Arguments.createMap();
                params.putString("utteranceId", utteranceId);
                params.putInt("start", start);
                params.putInt("end", end);
                params.putInt("frame", frame);
                sendEvent("onWordRange", params);
            }

            @Override
            public void onStart(String utteranceId) {
                WritableMap params = Arguments.createMap();
                params.putString("utteranceId", utteranceId);
                sendEvent("onSpeechStart", params);
            }

            @Override
            public void onDone(String utteranceId) {
                WritableMap params = Arguments.createMap();
                params.putString("utteranceId", utteranceId);
                sendEvent("onSpeechDone", params);
            }

            @Override
            public void onError(String utteranceId) {
                WritableMap params = Arguments.createMap();
                params.putString("utteranceId", utteranceId);
                sendEvent("onSpeechError", params);
            }
        });
    }

    private void sendEvent(String eventName, WritableMap params) {
        if (reactContext.hasActiveCatalystInstance()) {
            reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit(eventName, params);
        }
    }

    private WritableMap createErrorMap(String message) {
        WritableMap map = Arguments.createMap();
        map.putString("error", message);
        return map;
    }

    @ReactMethod
    public void speak(String text, ReadableMap params, Promise promise) {
        if (!isTtsReady) {
            promise.reject("TTS_NOT_READY", "TTS is not initialized yet");
            return;
        }

        String utteranceId = params.hasKey("utteranceId") ? params.getString("utteranceId")
                : String.valueOf(System.currentTimeMillis());
        Bundle bundle = new Bundle();
        bundle.putString(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, utteranceId);

        if (params.hasKey("rate")) {
            tts.setSpeechRate((float) params.getDouble("rate"));
        }
        if (params.hasKey("pitch")) {
            tts.setPitch((float) params.getDouble("pitch"));
        }
        if (params.hasKey("voice")) {
            String voiceId = params.getString("voice");
            setVoice(voiceId);
        }

        int result = tts.speak(text, TextToSpeech.QUEUE_FLUSH, bundle, utteranceId);
        if (result == TextToSpeech.SUCCESS) {
            promise.resolve(utteranceId);
        } else {
            promise.reject("TTS_ERROR", "Failed to speak");
        }
    }

    @ReactMethod
    public void stop(Promise promise) {
        if (tts != null) {
            tts.stop();
            promise.resolve(true);
        } else {
            promise.reject("TTS_NOT_READY", "TTS is not initialized");
        }
    }

    @ReactMethod
    public void pause(Promise promise) {
        // Android TTS doesn't support pause natively in the same way, usually stop is
        // used.
        // But we can just stop for now.
        stop(promise);
    }

    @ReactMethod
    public void getVoices(Promise promise) {
        if (!isTtsReady) {
            promise.reject("TTS_NOT_READY", "TTS is not initialized");
            return;
        }

        try {
            Set<Voice> voices = tts.getVoices();
            WritableArray voiceArray = Arguments.createArray();

            for (Voice voice : voices) {
                WritableMap voiceMap = Arguments.createMap();
                voiceMap.putString("identifier", voice.getName());
                voiceMap.putString("name", voice.getName());
                voiceMap.putString("language", voice.getLocale().toLanguageTag());
                voiceMap.putString("quality", voice.getQuality() + "");
                voiceArray.pushMap(voiceMap);
            }
            promise.resolve(voiceArray);
        } catch (Exception e) {
            promise.reject("GET_VOICES_ERROR", e.getMessage());
        }
    }

    private void setVoice(String voiceId) {
        if (tts == null || voiceId == null)
            return;
        try {
            for (Voice voice : tts.getVoices()) {
                if (voice.getName().equals(voiceId)) {
                    tts.setVoice(voice);
                    break;
                }
            }
        } catch (Exception e) {
            // Ignore if voice setting fails
        }
    }

    @ReactMethod
    public void addListener(String eventName) {
        // Keep: Required for RN built-in Event Emitter Calls.
    }

    @ReactMethod
    public void removeListeners(Integer count) {
        // Keep: Required for RN built-in Event Emitter Calls.
    }

    @Override
    public void onCatalystInstanceDestroy() {
        if (tts != null) {
            tts.stop();
            tts.shutdown();
        }
    }
}

File: android/app/src/main/java/com/rajarsheechatterjee/LNReader/TTSHighlightModule.kt
package com.rajarsheechatterjee.LNReader

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule


class TTSHighlightModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext), TTSForegroundService.TTSListener {

    private var ttsService: TTSForegroundService? = null
    private var isBound = false

    private val connection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            val binder = service as TTSForegroundService.TTSBinder
            ttsService = binder.getService()
            ttsService?.setTTSListener(this@TTSHighlightModule)
            isBound = true
        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            isBound = false
            ttsService = null
        }
    }

    init {
        val intent = Intent(reactContext, TTSForegroundService::class.java)
        reactContext.bindService(intent, connection, Context.BIND_AUTO_CREATE)
    }

    override fun getName(): String {
        return "TTSHighlight"
    }

    @ReactMethod
    fun speak(text: String, params: ReadableMap, promise: Promise) {
        if (!isBound || ttsService == null) {
            // Try to rebind
            val intent = Intent(reactContext, TTSForegroundService::class.java)
            reactContext.bindService(intent, connection, Context.BIND_AUTO_CREATE)
            promise.reject("TTS_NOT_READY", "TTS Service is not bound yet. Retrying...")
            return
        }

        val utteranceId = if (params.hasKey("utteranceId")) params.getString("utteranceId")!!
        else System.currentTimeMillis().toString()

        val rate = if (params.hasKey("rate")) params.getDouble("rate").toFloat() else 1.0f
        val pitch = if (params.hasKey("pitch")) params.getDouble("pitch").toFloat() else 1.0f
        val voiceId = if (params.hasKey("voice")) params.getString("voice") else null

        val success = ttsService?.speak(text, utteranceId, rate, pitch, voiceId) ?: false
        
        if (success) {
            promise.resolve(utteranceId)
        } else {
            promise.reject("TTS_ERROR", "Failed to start TTS. Service might not be ready.")
        }
    }
    
    @ReactMethod
    fun speakBatch(texts: ReadableArray, utteranceIds: ReadableArray, params: ReadableMap, promise: Promise) {
        if (!isBound || ttsService == null) {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
            return
        }
        
        val rate = if (params.hasKey("rate")) params.getDouble("rate").toFloat() else 1.0f
        val pitch = if (params.hasKey("pitch")) params.getDouble("pitch").toFloat() else 1.0f
        val voiceId = if (params.hasKey("voice")) params.getString("voice") else null
        
        val textList = mutableListOf<String>()
        val utteranceIdList = mutableListOf<String>()
        
        for (i in 0 until texts.size()) {
            texts.getString(i)?.let { text ->
                val utteranceId = utteranceIds.getString(i) ?: "utterance_$i"
                textList.add(text)
                utteranceIdList.add(utteranceId)
            }
        }
        
        val success = ttsService?.speakBatch(textList, utteranceIdList, rate, pitch, voiceId) ?: false
        
        if (success) {
            promise.resolve(textList.size)
        } else {
            promise.reject("TTS_ERROR", "Failed to start batch TTS")
        }
    }
    
    @ReactMethod
    fun addToBatch(texts: ReadableArray, utteranceIds: ReadableArray, promise: Promise) {
        if (!isBound || ttsService == null) {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
            return
        }
        
        val textList = mutableListOf<String>()
        val utteranceIdList = mutableListOf<String>()
        
        for (i in 0 until texts.size()) {
            texts.getString(i)?.let { text ->
                val utteranceId = utteranceIds.getString(i) ?: "utterance_$i"
                textList.add(text)
                utteranceIdList.add(utteranceId)
            }
        }
        
        val success = ttsService?.addToBatch(textList, utteranceIdList) ?: false
        
        if (success) {
            promise.resolve(true)
        } else {
            promise.reject("TTS_ERROR", "Failed to add to batch")
        }
    }
    
    @ReactMethod
    fun getQueueSize(promise: Promise) {
        if (isBound && ttsService != null) {
            promise.resolve(ttsService?.getQueueSize() ?: 0)
        } else {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
        }
    }

    @ReactMethod
    fun stop(promise: Promise) {
        if (isBound && ttsService != null) {
            ttsService?.stopTTS()
            promise.resolve(true)
        } else {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
        }
    }

    @ReactMethod
    fun pause(promise: Promise) {
        android.util.Log.d("TTS_DEBUG", "TTSHighlightModule.pause called")
        if (isBound && ttsService != null) {
            ttsService?.stopAudioKeepService()
            promise.resolve(true)
        } else {
            android.util.Log.e("TTS_DEBUG", "TTSHighlightModule.pause failed: Not bound")
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
        }
    }

    @ReactMethod
    fun updateMediaState(state: ReadableMap, promise: Promise) {
        if (!isBound || ttsService == null) {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
            return
        }

        val novelName = if (state.hasKey("novelName")) state.getString("novelName") else null
        val chapterLabel = if (state.hasKey("chapterLabel")) state.getString("chapterLabel") else null
        val chapterId = if (state.hasKey("chapterId") && !state.isNull("chapterId")) state.getInt("chapterId") else null
        val paragraphIndex = if (state.hasKey("paragraphIndex")) state.getInt("paragraphIndex") else 0
        val totalParagraphs = if (state.hasKey("totalParagraphs")) state.getInt("totalParagraphs") else 0
        val isPlaying = if (state.hasKey("isPlaying")) state.getBoolean("isPlaying") else false

        ttsService?.updateMediaState(
            novelName,
            chapterLabel,
            chapterId,
            paragraphIndex,
            totalParagraphs,
            isPlaying
        )

        promise.resolve(true)
    }

    @ReactMethod
    fun getVoices(promise: Promise) {
        if (isBound && ttsService != null) {
            try {
                val voices = ttsService?.getVoices() ?: emptyList()
                val voiceArray = Arguments.createArray()

                for (voice in voices) {
                    val voiceMap = Arguments.createMap()
                    voiceMap.putString("identifier", voice.name)
                    voiceMap.putString("name", voice.name)
                    voiceMap.putString("language", voice.locale.toLanguageTag())
                    voiceMap.putString("quality", voice.quality.toString())
                    voiceArray.pushMap(voiceMap)
                }
                promise.resolve(voiceArray)
            } catch (e: Exception) {
                promise.reject("GET_VOICES_ERROR", e.message)
            }
        } else {
            promise.reject("TTS_NOT_READY", "TTS Service is not bound")
        }
    }


    @ReactMethod
    fun addListener(eventName: String) {
        // Keep: Required for RN built-in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Keep: Required for RN built-in Event Emitter Calls.
    }

    // TTSListener implementation
    override fun onSpeechStart(utteranceId: String) {
        val params = Arguments.createMap()
        params.putString("utteranceId", utteranceId)
        sendEvent("onSpeechStart", params)
    }

    override fun onSpeechDone(utteranceId: String) {
        // Skip internal save position signals from TTSForegroundService
        // Position saving is now centralized in the Service
        if (utteranceId == TTSForegroundService.INTERNAL_SAVE_POSITION_SIGNAL) {
            return
        }
        
        val params = Arguments.createMap()
        params.putString("utteranceId", utteranceId)
        sendEvent("onSpeechDone", params)
    }

    override fun onSpeechError(utteranceId: String) {
        val params = Arguments.createMap()
        params.putString("utteranceId", utteranceId)
        sendEvent("onSpeechError", params)
    }

    override fun onWordRange(utteranceId: String, start: Int, end: Int, frame: Int) {
        val params = Arguments.createMap()
        params.putString("utteranceId", utteranceId)
        params.putInt("start", start)
        params.putInt("end", end)
        params.putInt("frame", frame)
        sendEvent("onWordRange", params)
    }

    override fun onQueueEmpty() {
        val params = Arguments.createMap()
        sendEvent("onQueueEmpty", params)
    }

    // FIX Case 7.2: Notify JS when voice fallback occurs
    override fun onVoiceFallback(originalVoice: String, fallbackVoice: String) {
        val params = Arguments.createMap()
        params.putString("originalVoice", originalVoice)
        params.putString("fallbackVoice", fallbackVoice)
        sendEvent("onVoiceFallback", params)
    }

    override fun onMediaAction(action: String) {
        android.util.Log.d("TTS_DEBUG", "TTSHighlightModule.onMediaAction received: $action")
        val params = Arguments.createMap()
        params.putString("action", action)
        sendEvent("onMediaAction", params)
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        android.util.Log.d("TTS_DEBUG", "TTSHighlightModule.sendEvent event=$eventName active=${reactContext.hasActiveCatalystInstance()}")
        @Suppress("DEPRECATION")
        if (reactContext.hasActiveCatalystInstance()) {
            try {
                reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit(eventName, params)
                android.util.Log.d("TTS_DEBUG", "TTSHighlightModule.sendEvent EMITTED")
            } catch (e: Exception) {
                android.util.Log.e("TTS_DEBUG", "TTSHighlightModule.sendEvent FAILED: ${e.message}")
            }
        } else {
             android.util.Log.w("TTS_DEBUG", "TTSHighlightModule.sendEvent SKIPPED (no active catalyst)")
        }
    }

    @Suppress("DEPRECATION")
    override fun onCatalystInstanceDestroy() {
        if (isBound) {
            reactContext.unbindService(connection)
            isBound = false
        }
        super.onCatalystInstanceDestroy()
    }
}
